### PR TITLE
overlord/state: introduce Change and NewChange on state to create them

### PR DESCRIPTION
**Pull request by pedronis**
This introduces Change objects under state as per spec, to track high level changes/intentions; Change comes with basic properties and Set/Get for entries support, and marshalling support.

This also adds State.NewChange to create Change objects.

SetStatus etc will come with a later branch.

### CODE REVIEW

File: overlord/state/change.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+type marshalledChange struct {
+	ID      string                      `json:"id"`
+	Kind    string                      `json:"kind"`
+	Summary string                      `json:"summary"`
+	Entries map[string]*json.RawMessage `json:"entries"`
+}
+
+// MarshalJSON makes Change a json.Marshaller
+func (c *Change) MarshalJSON() ([]byte, error) {
```
Comment (niemeyer) [inline overlord/state/change.go]: ensureLocked, just in case?  It needs to be locked.

File: overlord/state/change.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	return json.Marshal(marshalledChange{
+		ID:      c.id,
+		Kind:    c.kind,
+		Summary: c.summary,
+		Entries: c.entries,
+	})
+}
+
+// UnmarshalJSON makes Change a json.Unmarshaller
+func (c *Change) UnmarshalJSON(data []byte) error {
```
Comment (niemeyer) [inline overlord/state/change.go]: Same.

File: overlord/state/state.go [hunk 3]
```diff
@@ -37,6 +38,29 @@ type Backend interface {
 	Checkpoint(data []byte) error
 }
 
+type customEntries map[string]*json.RawMessage
```
Comment (niemeyer) [inline overlord/state/state.go]: I have a tempting feeling to name all of these as "data", "Data", and "customData", per the undefined nature of the contents. What do you think?

File: overlord/state/state.go [hunk 4]
```diff
@@ -51,19 +75,50 @@ type State struct {
...
 	}
 }
 
+type marshalledState struct {
+	Entries map[string]*json.RawMessage `json:"entries"`
+	Changes map[string]*Change          `json:"changes"`
+}
+
+// MarshalJSON makes State a json.Marshaller
+func (s *State) MarshalJSON() ([]byte, error) {
```
Comment (niemeyer) [inline overlord/state/state.go]: ensureLocked

File: overlord/state/state.go [hunk 5]
```diff
@@ -51,19 +75,50 @@ type State struct {
...
+	return json.Marshal(marshalledState{
+		Entries: s.entries,
+		Changes: s.changes,
+	})
+}
+
+// UnmarshalJSON makes State a json.Unmarshaller
+func (s *State) UnmarshalJSON(data []byte) error {
```
Comment (niemeyer) [inline overlord/state/state.go]: ensureLocked

File: overlord/state/state.go [hunk 6]
```diff
@@ -120,36 +175,55 @@ var ErrNoState = errors.New("no state entry for key")
...
 	s.ensureLocked()
-	serialized, err := json.Marshal(value)
-	if err != nil {
-		logger.Panicf("internal error: could not marshal value for state entry %q: %v", key, err)
+	s.entries.set(key, value)
+}
+
+var rnd = rand.New(rand.NewSource(time.Now().UTC().UnixNano()))
+
+func (s *State) genID() string {
```
Comment (niemeyer) [inline overlord/state/state.go]: strutils.MakeRandomString should work for this

**Comment by niemeyer** [comment]
LGTM, with just a few trivials.

**Comment by zyga** [comment]
I've reviewed the whole delta and I cannot see anything beyond what Gustavo already suggested. +1 and I cannot wait to use it :-)

**Pull request merged**